### DISCUSSION TITLE
Web sockets implementation

### DISCUSSION THREAD

**Discussion by kasun** [zauberzeug/nicegui#728] [Q&A] [answered]
First of all thanks for putting up nicegui.

I'm interested in knowing how the web socket connections are implemented. Is a web socket connection opened when a user visit the app/a page and remained opened until the page is closed? Or is a web socket opened for each event and closed after handling the event?

**Comment by rodja** [comment, marked as answer]
Thank you for your interest in NiceGUI and it's intrinsics @kasun. We use the socket.io library for managing web sockets. After the initial content is loaded a web socket connection is established and kept open for communication as long as the web page is shown. Each http request gets its own web socket connection.

**Comment by kasun** [comment]
Thanks for the explanation! Have you done any performance testing on a nicegui app? What kind of traffic does nicegui.io site handle? I'm interested in knowing how scalable this approach can be.

**Reply by rodja** [reply]
We have not done any formal performance tests. But we would welcome any efforts in this direction. NiceGUI needs to send all user interactions from the frontend to the backend and ui updates back to the server via the web socket. We have taken much care to reduce the data load. But optimizations never end. A thorough performance analysis would certainly help.
